File: mathlearn/backend/app/schemas/stats.py
"""Схемы для статистики."""

from datetime import datetime
from pydantic import BaseModel


class HeatmapCell(BaseModel):
    """Схема ячейки heatmap."""

    factor_a: int
    factor_b: int
    error_count: int = 0
    avg_time_ms: int = 0
    accuracy: float = 0.0


class HeatmapResponse(BaseModel):
    """Схема ответа с heatmap ошибок (матрица 10x10)."""

    matrix: list[list[HeatmapCell]]


class SpeedDataPoint(BaseModel):
    """Схема точки данных скорости."""

    date: datetime
    avg_response_ms: int
    accuracy: float


class SpeedResponse(BaseModel):
    """Схема ответа с динамикой скорости ответов."""

    data_points: list[SpeedDataPoint]
    days: int


class StreakResponse(BaseModel):
    """Схема ответа с данными о streak."""

    current_streak: int
    max_streak: int


class AchievementResponse(BaseModel):
    """Схема достижения."""

    id: int
    name: str
    description: str
    unlocked: bool
    unlocked_at: datetime | None


class AchievementsResponse(BaseModel):
    """Схема ответа со списком достижений."""

    achievements: list[AchievementResponse]

File: mathlearn/frontend/src/api/stats.ts
import apiClient from './client';

export interface HeatmapCell {
  factor_a: number;
  factor_b: number;
  error_count: number;
  avg_time_ms: number;
  accuracy: number;
}

export interface HeatmapResponse {
  matrix: HeatmapCell[][];
}

export interface SpeedDataPoint {
  date: string;
  avg_response_ms: number;
  accuracy: number;
}

export interface SpeedResponse {
  data_points: SpeedDataPoint[];
  days: number;
}

export interface StreakResponse {
  current_streak: number;
  max_streak: number;
}

export interface AchievementResponse {
  id: number;
  name: string;
  description: string;
  unlocked: boolean;
  unlocked_at: string | null;
}

export interface AchievementsResponse {
  achievements: AchievementResponse[];
}

export const statsApi = {
  getHeatmap: async (): Promise<HeatmapResponse> => {
    const response = await apiClient.get('/stats/heatmap');
    return response.data;
  },

  getSpeed: async (days: number = 30): Promise<SpeedResponse> => {
    const response = await apiClient.get(`/stats/speed?days=${days}`);
    return response.data;
  },

  getStreak: async (): Promise<StreakResponse> => {
    const response = await apiClient.get('/stats/streak');
    return response.data;
  },

  getAchievements: async (): Promise<AchievementsResponse> => {
    const response = await apiClient.get('/stats/achievements');
    return response.data;
  },
};
